drivers/ata/ahciem.c

```diff
@@ -0,0 +1,566 @@
+		int offset = 4 + 4 + (4 + 8) * i; /* pghdr + gencode + slots */
+
+		/* Additional Element Status Descriptor */
+		rbuf[offset] = 0x10 | SCSI_PROTOCOL_ATA;	/* eip=1, proto=ATA */
+		rbuf[offset + 1] = 2 + 8;	/* length: index + ata elm */
+		rbuf[offset + 2] = 0x01;	/* eiioe */
+		rbuf[offset + 3] = 1 + i;	/* index */
+
+		ap = host->ports[i];
+		if (sata_pmp_attached(ap))
+			rbuf[offset] |= 0x80;	/* invalid */
+
+		/* ATA Element Status (NB: non-standard) */
+		put_unaligned_be32(i, rbuf + offset + 4);
+		put_unaligned_be32(ap->scsi_host->host_no, rbuf + offset + 8);
+	}
+
+	return 0;
+}
+
+static void ahciem_setleds(struct ahciem_enclosure *enc, int slot)
+{
+	struct ata_port *ap = enc->host->ports[slot];
+	u32 port_led_state, val;
+
+	if (!ap->ops->transmit_led_message)
+		return;
+
+	val = 0;
+	if (enc->status[slot][2] & 0x80)		/* Activity */
+		val |= (1 << 0);
+	if (enc->status[slot][1] & 0x02)		/* Rebuild */
+		val |= (1 << 6) | (1 << 3);
+	else if (enc->status[slot][2] & 0x02)		/* Identification */
+		val |= (1 << 3);
+	else if (enc->status[slot][3] & 0x20)		/* Fault */
+		val |= (1 << 6);
+
+	port_led_state = (val << 16) | (slot & EM_MSG_LED_HBA_PORT);
+
+	ap->ops->transmit_led_message(ap, port_led_state, 4);
+}
+
+static void ahciem_sesop_txdx(struct ahciem_enclosure *enc, struct scsi_cmnd *cmd)
+{
+	const u8 *ads0;
+	u8 *page;
+	u16 page_len = get_unaligned_be16(cmd->cmnd + 3);
+	int n_ports = enc->host->n_ports;
+	int i;
+
+	page = kzalloc(page_len, GFP_KERNEL);
+	if (!page) {
+		ahciem_scsi_set_sense(cmd, ABORTED_COMMAND, 0x34, 0x0);
+		return;
+	}
+
+	if (scsi_sg_copy_to_buffer(cmd, page, page_len) != page_len) {
+		kfree(page);
+		ahciem_scsi_set_sense(cmd, ABORTED_COMMAND, 0x34, 0x0);
+		return;
+	}
+
+	if (page[0] != 0x02) {	/* Enclosure Control page */
+		kfree(page);
+		ahciem_scsi_set_invalid_field(cmd, 0, 0, false);
+		return;
+	}
+
+	ads0 = page + 8;
+
+	for (i = 0; i < n_ports; i++) {
+		const u8 *ads = ads0 + 4 + 4 * i; /* start + overall elem + elems */
+
+		if (ads[0] & 0x80) {
+			enc->status[i][0] = 0;
+			enc->status[i][1] = ads[1] & 0x02;		/* rebuild/remap */
+			enc->status[i][2] = ads[2] & (0x80 | 0x02);	/* rqst active + rqst ident */
+			enc->status[i][3] = ads[3] & 0x20;		/* rqst fault */
+			ahciem_setleds(enc, i);
+		} else if (ads0[0] & 0x80) {
+			enc->status[i][0] = 0;
+			enc->status[i][1] = ads0[1] & 0x02;		/* rebuild/remap */
+			enc->status[i][2] = ads0[2] & (0x80 | 0x02);	/* rqst active + rqst ident */
+			enc->status[i][3] = ads0[3] & 0x20;		/* rqst fault */
+			ahciem_setleds(enc, i);
+		}
+	}
+
+	kfree(page);
+}
+
+static int ahciem_queuecommand(struct Scsi_Host *shost, struct scsi_cmnd *cmd)
+{
+	struct ahciem_enclosure *enc = (struct ahciem_enclosure *)&shost->hostdata[0];
+	struct ahciem_args args = { .cmd = cmd, .enc = enc };
+	const u8 *cdb = cmd->cmnd;
+
+	if (unlikely(!cmd->cmd_len)) {
+		printk("ahciem: bad CDB len=0 in %s\n", DRV_NAME);
+		cmd->result = DID_ERROR << 16;
+		scsi_done(cmd);
```

Comment:
This line was modified. Previously, `scsi_cmnd` struct contained a function pointer for scsi_done callback. But now it is decalared as a function in scsi.h.